File: trabalho2-elixir/solver.exs
defmodule Matrix do
  @moduledoc """
    Módulo com métodos auxiliares para problemas com matrizes.
    Inclui a criação de tipos de dados que encapsulam listas e valores inteiros.
  """

  @typedoc """
    Tipo valor de cada posição (encapsula inteiros)
  """
  @type value_t :: integer

  @typedoc """
    Linhas das matrizes são listas de valores
  """
  @type row_t :: [value_t]

  @typedoc """
    Define o tipo Matrix, com linhas de um determinado tipo
  """
  @type matrix_t :: [row_t]

  @typedoc """
    Define um Table, que é matriz de int
  """
  @type table_t :: matrix_t

  @typedoc """
    Lista de valores possiveis para uma determinada célula
  """
  @type choices_t :: [value_t]

  ##################################################
  # definição dos métodos do modulo

  @doc """
    Retorna tamanho da matriz
  """
  @spec size(matrix_t) :: integer
  def size(matrix), do: length(Enum.at(matrix, 1))

  @doc """
    Retorna linhas da matriz em lista
  """
  @spec rows(matrix_t) :: [row_t]
  def rows(matrix), do: matrix

  @doc """
    Retorna colunas da matriz em lista
  """
  @spec cols(matrix_t) :: [row_t]
  def cols(matrix) do
    matrix
    |> List.zip()
    # & é o operador usado para transformar função em função anonima
    |> Enum.map(&Tuple.to_list/1)
  end

  @doc """
    cria tupla com valor e grupo de determinada posição
    divide a matriz de valores de acordo com os grupos
    retorna matriz com blocos
  """
  @spec matrixByGroup(matrix_t, table_t) :: matrix_t
  def matrixByGroup(values, groups) do
    for group <- mapGroups(valuesTuple(values, groups)),
        do: filterByGroup(group, valuesTuple(values, groups))
  end

  @doc """
    cria tupla (valor, grupo) para cada posição do tabuleiro
  """
  @spec valuesTuple(matrix_t, table_t) :: matrix_t
  def valuesTuple(matrix1, matrix2) do
    Enum.zip(matrix1, matrix2)
    |> Enum.flat_map(fn {row1, row2} -> Enum.zip(row1, row2) end)
  end

  # recebe lista de tuplas
  defp mapGroups(valuesG) do
    Enum.map(valuesG, fn {_, x} -> x end) |> Enum.uniq()
  end

  defp filterByGroup(group, valueTuple) do
    Enum.filter(valueTuple, fn {_, g} -> g == group end)
    |> Enum.map(fn {f, _} -> f end)
  end

  @doc """
    A partir da lista de linhas, remonta a matriz original (recebe separada por grupos)
  """
  @spec originalCols([row_t], integer) :: [row_t]
  def originalCols(bs, n) do
    bs
    |> Enum.concat()
    |> divide_list(n)
  end

  defp divide_list([], _n), do: []

  defp divide_list(list, n) do
    [Enum.take(list, n) | divide_list(Enum.drop(list, n), n)]
  end

  @doc """
    A partir dos valores e grupos, gera a matriz separada em grupos
  """
  @spec groupsByColumn(matrix_t, table_t) :: [row_t]
  def groupsByColumn(values, groups) do
    Enum.zip(Matrix.cols(values), Matrix.cols(groups))
    |> Enum.reduce([], fn {column1, column2}, acc ->
      grouped =
        Enum.with_index(Enum.zip(column1, column2))
        |> Enum.group_by(fn {{_, y}, _} -> y end)
        |> Enum.map(fn {_, list} -> {hd(list), list} end)
        |> Enum.sort_by(fn {{_, n}, _} -> n end)
        |> Enum.map(fn {_, value} -> value end)
        |> Enum.map(fn sublist -> Enum.map(sublist, fn {{number, _}, _} -> number end) end)

      acc ++ grouped
    end)
  end
end

defmodule Solver do
  @moduledoc """
    Módulo com funções responsáveis pela solução de tabuleiros do jogo Kojun
  """

  @doc """
    retorna o tamanho de um determinado grupo (dado pelo id)
  """
  @spec groupSize(integer, Matrix.matrix_t()) :: integer
  def groupSize(id, groups) do
    Enum.sum(for group <- groups, do: Enum.count(group, &(&1 == id)))
  end

  @doc """
    Retorna os valores que estão em um grupo especificado pelo seu id de grupo
  """
  @spec getValuesInGroup(Matrix.matrix_t(), Matrix.table_t(), integer) :: [integer]
  def getValuesInGroup(values, groups, id) do
    Enum.map(Enum.filter(Matrix.valuesTuple(values, groups), fn {_, v} -> v == id end), fn {v, _} ->
      v
    end)
  end

  @doc """
    Retorna a matriz de escolhas a partir dos valores e grupos
  """
  @spec choices(Matrix.table_t(), Matrix.table_t()) :: Matrix.matrix_t()
  def choices(values, groups) do
    Enum.map(Enum.zip(values, groups), fn {value, group} ->
      Enum.map(Enum.zip(value, group), fn {v, g} ->
        if v == 0 do
          Enum.to_list(1..groupSize(g, groups)) -- getValuesInGroup(values, groups, g)
        else
          [v]
        end
      end)
    end)
  end

  @doc """
    Posições que só possui um valor possível, seta o valor na célula
  """
  @spec reduceChoicesByList(Matrix.row_t()) :: Matrix.row_t()
  def reduceChoicesByList(xss) do
    singles = xss |> Enum.filter(&Solver.singleElementInList/1) |> List.flatten()
    Enum.map(xss, &Solver.minus(&1, singles))
  end

  @doc """
    Faz a redução das escolhas dentro de cada lista e retorna a matriz reduzida
  """
  @spec reduceChoices(Matrix.matrix_t(), Matrix.table_t()) :: Matrix.matrix_t()
  def reduceChoices(values, groups) do
    groupsByCol = Matrix.groupsByColumn(values, groups)
    reduction = Enum.map(groupsByCol, &reduceChoicesByList/1)
    Matrix.originalCols(reduction, Matrix.size(values)) |> Matrix.cols()
  end

  @doc """
    Ordenamento descrescente de cima para baixo nas colunas dentro de um grupo
  """
  @spec descendingRow(list) :: bool
  def descendingRow([]), do: true
  def descendingRow([_]), do: true

  def descendingRow([a, b | bs]) when length(a) <= 1 and length(b) <= 1 do
    if a < b, do: false, else: descendingRow([b | bs])
  end

  def descendingRow([_, b | bs]), do: descendingRow([b | bs])

  @doc """
    Compara valores iguais dentro de cada grupo
  """
  @spec validRow(list) :: bool
  def validRow([]), do: true

  def validRow([x | xs]) when length(x) <= 1 do
    not Enum.member?(xs, x) and validRow(xs)
  end

  def validRow([_ | xs]), do: validRow(xs)

  @doc """
    Nenhuma célula tem vizinhos com valor igual (em linha e coluna)
  """
  @spec validNeighbour(Matrix.row_t()) :: bool
  def validNeighbour([]), do: true
  def validNeighbour([_]), do: true

  def validNeighbour([a, b | bs]) when length(a) <= 1 and length(b) <= 1 do
    if a == b, do: false, else: validNeighbour([b | bs])
  end

  def validNeighbour([_, b | bs]), do: validNeighbour([b | bs])

  @doc """
    Verifica se há apenas um elemento na lista
  """
  @spec singleElementInList([Matrix.value_t()]) :: bool
  def singleElementInList([_]), do: true
  def singleElementInList(_), do: false

  @doc """
    Verifica se há elementos vazios e se as regras do jogo são cumpridas
  """
  @spec notPossible(Matrix.matrix_t(), Matrix.table_t()) :: bool
  def notPossible(values, groups) do
    empty =
      Enum.any?(values, fn row ->
        Enum.any?(row, fn r ->
          r == []
        end)
      end)

    empty || !valid(values, groups)
  end

  @doc """
    Testa todas regras do jogo
  """
  @spec valid(Matrix.matrix_t(), Matrix.table_t()) :: bool
  def valid(values, groups) do
    Enum.reduce(Matrix.cols(values), true, fn col, acc -> acc && validNeighbour(col) end) &&
      Enum.reduce(Matrix.rows(values), true, fn row, acc -> acc && validNeighbour(row) end) &&
      Enum.reduce(Matrix.matrixByGroup(values, groups), true, fn matrixGroup, acc ->
        acc && validRow(matrixGroup)
      end) &&
      Enum.reduce(Matrix.groupsByColumn(values, groups), true, fn groupColumn, acc ->
        acc && descendingRow(groupColumn)
      end)
  end

  @doc """
    Expansão da matriz de escolhas
  """
  @spec expandChoices(Matrix.matrix_t()) :: [Matrix.matrix_t()]
  def expandChoices(m) do
    {rows1, [row | rows2]} = Enum.split_while(m, fn r -> Enum.all?(r, &singleElementInList/1) end)
    {row1, [cs | row2]} = Enum.split_while(row, &singleElementInList/1)

    Enum.map(cs, fn c ->
      Enum.concat([rows1, [Enum.concat([row1, [[c]], row2])], rows2])
    end)
  end

  @doc """
    Remove os elementos que estão na lista xs de ys
  """
  @spec minus(Matrix.choices_t(), Matrix.choices_t()) :: Matrix.choices_t()
  def minus(xs, ys) do
    if singleElementInList(xs) do
      xs
    else
      xs -- ys
    end
  end

  @doc """
    Cria uma lista com tabuleiros com possíveis soluções para o tabuleiro e por meio do backtracking filtra as escolhas
  """
  @spec searchForSolution(Matrix.matrix_t(), Matrix.table_t()) :: [Matrix.table_t()]
  def searchForSolution(values, groups) do
    cond do
      Solver.notPossible(values, groups) ->
        []

      Enum.all?(values, fn sublist -> Enum.all?(sublist, &singleElementInList/1) end) ->
        [values]

      true ->
        for values1 <- expandChoices(values),
            g <- searchForSolution(reduceChoices(values1, groups), groups),
            do: g
    end
  end

  @doc """
    Encontra a solução para o tabuleiro (devolve a primeira encontrada)
  """
  @spec getSolution(Matrix.table_t(), Matrix.table_t()) :: Matrix.table_t()
  def getSolution(values, groups) do
    Enum.at(searchForSolution(reduceChoices(choices(values, groups), groups), groups), 0)
  end

  @doc """
    Função para imprimir a matriz final obtida na tela com formato correto
  """
  @spec convertOutput([Matrix.matrix_t()]) :: Matrix.matrix_t()
  def convertOutput(matrix) do
    for row <- matrix, do: for(el <- row, do: Enum.at(el, 0))
  end
end

values6x6 = [
  [2, 0, 0, 0, 1, 0],
  [0, 0, 0, 3, 0, 0],
  [0, 3, 0, 0, 5, 3],
  [0, 0, 0, 0, 0, 0],
  [0, 0, 3, 0, 4, 2],
  [0, 0, 0, 0, 0, 0]
]

groups6x6 = [
  [1, 1, 2, 2, 2, 3],
  [4, 4, 4, 4, 4, 3],
  [5, 6, 6, 6, 4, 7],
  [5, 5, 5, 6, 7, 7],
  [8, 8, 10, 0, 0, 0],
  [9, 9, 10, 10, 0, 0]
]

values8x8 = [
  [2, 5, 0, 0, 3, 0, 0, 0],
  [0, 0, 6, 0, 0, 0, 0, 0],
  [0, 0, 5, 0, 5, 2, 0, 0],
  [0, 0, 0, 2, 0, 0, 0, 0],
  [0, 0, 1, 0, 4, 0, 0, 0],
  [3, 0, 2, 0, 0, 4, 0, 0],
  [0, 0, 0, 6, 0, 0, 0, 0],
  [0, 0, 0, 0, 4, 0, 3, 2]
]

groups8x8 = [
  [0, 1, 1, 1, 1, 2, 3, 3],
  [0, 0, 5, 1, 2, 2, 4, 4],
  [6, 7, 5, 8, 2, 2, 9, 9],
  [6, 10, 5, 5, 5, 2, 9, 9],
  [6, 10, 5, 11, 11, 11, 9, 12],
  [13, 10, 14, 11, 14, 14, 12, 12],
  [13, 13, 14, 14, 14, 15, 15, 12],
  [13, 13, 16, 16, 15, 15, 15, 12]
]

solution = Solver.getSolution(values6x6, groups6x6)
Solver.convertOutput(solution) |> IO.inspect()
# IO.inspect(Solver.getSolution(values8x8, groups8x8))

File: trabalho1-haskell/solver.hs
import Data.List

-- definições dos tipos
type Value = Int
type Row a = [a]
type Matrix a = [Row a]
type Table = Matrix Value
type Choices = [Value]

------------------------------------------------------------------
-- métodos para operar as matrizes

-- retorna ordem da matriz (quadrada)
size :: Matrix m -> Int
size matrix = length (head matrix)

-- linhas de uma matriz
rows :: Matrix m -> [Row m]
rows matrix = matrix

-- colunas de uma matriz
cols :: Matrix m -> [Row m]
cols matrix = transpose matrix

-- cria tupla com valor e grupo de determinada posição
-- divide a matriz de valores de acordo com os grupos
-- retorna matriz com blocos
matrixByGroup :: Eq m => Matrix m -> Table -> Matrix m
matrixByGroup values groups = [filterByGroup group valuesTuple | group <- mapGroups]
    where
        valuesTuple = foldl1 (++) (zipWith zip values groups)
        mapGroups = nub (map snd valuesTuple)
        filterByGroup group list = map fst $ filter ((== group) . snd) list

-- remonta as colunas originais a partir de uma dividida por blocos
originalCols :: [Row l] -> Int -> [Row l]
originalCols bs n = divideList n (concat bs)
    where
        divideList n = takeWhile (not . null) . map (take n) . iterate (drop n)

------------------------------------------------------------------
-- métodos da solução do Kojun

-- conta quantos elementos estão em um grupo para saber seu tamanho
groupSize :: Eq m => m -> Matrix m -> Int
groupSize _ [] = 0
groupSize id groups = sum [count id p | p <- groups]
  where count x xs = length (filter (==x) xs)

-- id identifica um bloco (grupo)
-- cria lista dos valores que já estão em um bloco
getValuesInGroup :: Eq m => Matrix m -> Table -> Int -> [m]
getValuesInGroup values groups id = map fst $ filter ((==id) . snd) valuesTuple
    where
        valuesTuple = foldl1 (++) (zipWith zip values groups)

minus :: Choices -> Choices -> Choices
xs `minus` ys = if singleElementInList xs then xs else xs \\ ys

-- ve se a linha tem apenas um elemento
singleElementInList :: [a] -> Bool
singleElementInList [_] = True
singleElementInList _ = False

-- cria escolhas possíveis em cada célula
-- células sem valor possuem valores possíveis de 1 até tamanho do grupo
-- remove os valores que já estão naquele grupo (para não haver repetição)
choices :: Table -> Table -> Matrix Choices
choices values groups = map (map choice) (zipWith zip values groups)
    where choice (v, p) = if v == 0 then [1..(groupSize p groups)] `minus` (getValuesInGroup values groups p) else [v]

-- cria as colunas da matriz delimitando pelos grupos
groupsByColumn :: Eq m => Matrix m -> Table -> [Row m]
groupsByColumn values groups = zipWith zip (cols values) (cols groups) >>= map (map fst) . groupBy (\x y -> snd x == snd y)

-- em cada coluna (já dividida por grupos), faz uma redução em lista
-- faz a redução das escolhas dentro de cada lista e retorna a matriz reduzida
reduceChoices :: Matrix Choices -> Table -> Matrix Choices
reduceChoices values groups = cols $ originalCols (map reduceChoicesByList (groupsByColumn values groups)) (size values)

-- em posições que só possui um valor possível, seta o valor na célula
reduceChoicesByList :: Row Choices -> Row Choices
reduceChoicesByList xss = [xs `minus` singles | xs <- xss]
    where singles = concat (filter singleElementInList xss)

-- filtra soluções baseada nas escolhas, percorrendo as células
-- cria uma lista com tabuleiros com possíveis soluções para o tabuleiro
searchForSolution :: Matrix Choices -> Table -> [Table]
searchForSolution values groups
    | notPossible values groups = []  -- caso não tenha solução, retorna lista vazia
    | all (all singleElementInList) values = [map concat values]  -- não precisa de mais reduções
    | otherwise = [g | values' <- expandChoices values, g <- searchForSolution (reduceChoices values' groups) groups]  -- ainda precisa de redução, expandindo a matriz e em seguida reduzindo, continuando a busca

-- ve se tabuleiro não gera solução
notPossible :: Matrix Choices -> Table -> Bool
notPossible values groups = empty values || not (valid values groups)
    where
        empty m = any (any null) m

-- verifica se uma matriz de escolhas é valida
valid :: Matrix Choices -> Table -> Bool
valid values groups = all validNeighbour (cols values) &&
                      all validNeighbour (rows values) &&  -- nenhuma célula tem vizinhos com valor igual (em linha e coluna)
                      all validRow (matrixByGroup values groups) &&  --compara dentro do grupo para ver se não há valor igual
                      all descendingRow (groupsByColumn values groups)  -- deve haver ordenamento descrescente de cima para baixo nas colunas dentro de um grupo

-- vizinhos não podem ter o mesmo valor
validNeighbour :: Eq a => Row [a] -> Bool
validNeighbour [] = True
validNeighbour [a] = True
validNeighbour (a:b:bs)  -- compara primeiro com segundo, se necessáiro, chama a função usando o segundo valor e o resto da lista 
    | (length a <= 1) && (length b <= 1) = if a == b then False else validNeighbour (b:bs)
    | otherwise = validNeighbour (b:bs)  -- b = segundo valor; bs = restante da Row

-- não pode haver valores iguais em uma linha
validRow :: Eq a => Row [a] -> Bool
validRow [] = True
validRow (x:xs) = if (length x <= 1) then not (elem x xs) && validRow xs else validRow xs

-- ordenamento decrescente nas colunas (não pode ter valor menor sobre maior)
descendingRow :: Ord a => Row [a] -> Bool
descendingRow [] = True
descendingRow [a] = True
descendingRow (a:b:bs)
    | (length a <= 1) && (length b <= 1) = if a < b then False else descendingRow (b:bs)
    | otherwise = descendingRow (b:bs)

-- faz expanção das escolhas
expandChoices :: Matrix Choices -> [Matrix Choices]
expandChoices m = [rows1 ++ [row1 ++ [c] : row2] ++ rows2 | c <- cs]
    where
        (rows1,row:rows2) = break (not . all singleElementInList) m
        (row1,cs:row2) = break (not . singleElementInList) row

-- encontra a solução para o tabuleiro (devolve a primeira encontrada)
getSolution :: Table -> Table -> Table
getSolution values groups = head (searchForSolution (reduceChoices (choices values groups) groups) groups)

values6x6 :: Table
values6x6 = [[2,0,0,0,1,0],
             [0,0,0,3,0,0],
             [0,3,0,0,5,3],
             [0,0,0,0,0,0],
             [0,0,3,0,4,2],
             [0,0,0,0,0,0]
            ]

groups6x6 :: Table
groups6x6 = [[1,1,2,2,2,3],
             [4,4,4,4,4,3],
             [5,6,6,6,4,7],
             [5,5,5,6,7,7],
             [8,8,10,0,0,0],
             [9,9,10,10,0,0]
            ]

-- https://www.janko.at/Raetsel/Kojun/065.a.htm
-- values12x12 :: Table
-- values12x12 = [
--              [0,0,0,4,0,3,2,0,0,1,0,3],
--              [0,0,2,0,5,0,0,0,0,0,0,0],
--              [0,4,0,0,0,1,0,0,0,0,0,0],
--              [2,0,5,0,2,0,0,0,4,0,0,0],
--              [1,6,0,0,0,4,0,2,0,0,0,0],
--              [0,5,2,0,0,0,0,0,2,0,0,1],
--              [0,0,0,0,5,0,4,0,0,4,0,0],
--              [3,0,0,7,0,0,0,0,0,0,3,0],
--              [0,0,6,0,0,0,0,3,0,0,2,0],
--              [0,5,0,3,0,0,0,4,0,5,0,0],
--              [4,0,0,0,0,2,0,0,6,0,0,0],
--              [5,0,3,0,0,0,3,1,4,0,1,0]
--             ]
-- groups12x12 :: Table
-- groups12x12 = [
--     [0,0,1,1,2,2,2,2,3,4,4,5],
--     [6,6,1,7,7,7,8,9,3,3,5,5],
--     [10,10,1,11,7,7,8,12,3,5,5],
--     [10,11,11,11,11,]
-- ]

main = do
    let values = values6x6
    let groups = groups6x6
    print "Resolvendo tabuleiro 6x6"
    mapM_ print (getSolution values groups)